src/main/scala/edu/berkeley/ce/rockslicing/BlockVTK.scala

```diff
@@ -0,0 +1,197 @@
+        val rotatedVerts = kv._2 map { vertex =>
+          val rotatedVertex = R * DenseVector[Double](vertex._1, vertex._2, vertex._3)
+          (rotatedVertex(0), rotatedVertex(1), rotatedVertex(2))
+        }
+        // Order vertices in counter-clockwise orientation
+        val center = BlockVTK.findCenter(rotatedVerts)
+        val orderedVerts = {
+          if (kv._1.normalVec._3 < -BlockVTK.EPSILON) {
+            rotatedVerts.sortWith(BlockVTK.ccwCompare(_, _, center)).reverse
+          }
+          else {
+            rotatedVerts.sortWith(BlockVTK.ccwCompare(_, _, center))
+          }
+        }
+        // Rotate vertices back to original orientation
+        val invR = R.t // Inverse of rotation matrix is equal to its transpose
+        orderedVerts map { vertex =>
+          val orderedVertex = invR * DenseVector[Double](vertex._1, vertex._2, vertex._3)
+          (orderedVertex(0), orderedVertex(1), orderedVertex(2))
+        }
+      }
+    ).toMap
+  }
+
+  /**
+   * Finds the indices of the face vertices in the global list of vertices
+   * @param faceOrientedVerts Mapping from each face to a Seq of vertices for that face
+   * @param vertices Seq of distinct vertices for the rock block
+   * @return A mapping from each face of the block to a Seq of integers that represent
+   *         the indices of the vertices in the global vertex list
+   */
+  private def connectivity(faceOrientedVerts: Map[Face, Seq[(Double, Double, Double)]],
+                           vertices: Seq[(Double, Double, Double)]): Map[Face, Seq[Int]] = {
+    faceOrientedVerts.keys.zip(
+      faceOrientedVerts map { kv =>
+        kv._2 map { case vertex =>
+          vertices.indexOf(vertex)
+        }
+      }
+    ).toMap
+  }
+
+  /**
+   * Creates a sequence of all the normals of all the faces for the list of input blocks
+   * @param faces A mapping from each face of the block to a Seq of vertices for that face arranged in counter-
+   *         clockwise order relative to its unit normal
+   * @return Sequence of tuples representing normals of all the block faces
+   */
+  private def normals(faces: Map[Face, Seq[(Double, Double, Double)]]): Seq[(Double, Double, Double)] = {
+    faces map { kv =>
+      (kv._1.a, kv._1.b, kv._1.c)
+    }
+  }.toSeq
+
+  /**
+   * Determines the cumulative offsets that define each face in the connectivity list
+   * @param connectivities Mapping from each face of a block to a Seq of integers that represent
+   *                       tne indices of the vertices in the global vertex list
+   * @return A Seq of integers that represent the cumulative offset of each face in the
+   *         connectivity list
+   */
+  private def faceOffsets(connectivities: Map[Face, Seq[Int]]): Seq[Int] = {
+    val localOffsets = (connectivities map { kv => kv._2.length}).toList
+    var offsets = List[Int](localOffsets.head)
+    for (i <- 1 until localOffsets.length) {
```

Comment:
I can try. Admittedly, I was struggling to find a functional way to deal with this since each entry depends on the previous one. Do you have any suggestions?
Also, what would you suggest to overcome the issues with adding to the end of `List`?